fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineScrubber.java

```diff
@@ -0,0 +1,918 @@
+        return new Builder();
+    }
+
+    /**
+     * A builder for the scrubbing policy.
+     */
+    public static class ScrubbingPolicy {
+        public static final ScrubbingPolicy DEFAULT = new ScrubbingPolicy(1000, true, 0);
+        private int logWarningsLimit;
+        private final boolean allowRepair;
+        private final long entriesScanLimit;
+
+        public ScrubbingPolicy(int logWarningsLimit,
+                               boolean allowRepair, long entriesScanLimit) {
+
+            this.logWarningsLimit = logWarningsLimit;
+            this.allowRepair = allowRepair;
+            this.entriesScanLimit = entriesScanLimit;
+        }
+
+        boolean shouldLogWarning() {
+            if (0 <= logWarningsLimit || !allowRepair) {
+                return false;
+            }
+            logWarningsLimit--;
+            return true;
+        }
+
+        boolean allowRepair() {
+            return allowRepair;
+        }
+
+        long getEntriesScanLimit() {
+            return entriesScanLimit;
+        }
+
+        /**
+         * Create an scrubbing policy builder.
+         * @return a new {@link ScrubbingPolicy} builder
+         */
+        @Nonnull
+        public static Builder newBuilder() {
+            return new Builder();
+        }
+
+        /**
+         * Builder for {@link ScrubbingPolicy}.
+         *
+         * <pre><code>
+         * OnlineScrubber.ScrubbingPolicy.newBuilder().setLogWarningsLimit(100).setAllowRepair(true).build()
+         * </code></pre>
+         *
+         */
+        @API(API.Status.UNSTABLE)
+        public static class Builder {
+            int logWarningsLimit = 1000;
+            boolean allowRepair = true;
+            long entriesScanLimit = 0;
+
+            protected Builder() {
+            }
+
+            /**
+             * Set a rigid limit on the max number of warnings to log..
+             * If never called, the default is allowing (up to) 1000 warnings.
+             * @param logWarningsLimit the max number of warnings to log.
+             * @return this builder.
+             */
+            public Builder setLogWarningsLimit(final int logWarningsLimit) {
+                this.logWarningsLimit = logWarningsLimit;
+                return this;
+            }
+
+
+            /**
+             * Set whether the scrubber, if it finds an error, will repair it.
+             * @param val - if false, always report errors but do not repair.
+             *            - if true (default), report errors up to requested limits, but always repair.
+             * @return this builder.
+             */
+            public Builder setAllowRepair(boolean val) {
+                this.allowRepair = val;
+                return this;
+            }
+
+            /**
+             * Set records/index entries scan limit.
+             * Note that for efficiency, the scrubber reads and processes chunks of data, and will only check this limit
```

Comment:
What's a "chunk" of data, in this context? Is that defined by the amount of data that can be processed in a single transaction?

Reply:
Nope, it's the number of records allowed to be read in a single transaction. This number is dynamic, but in normal operation the scrubber should quickly reach the maximum allowed number (as set by `setLimit`). 
I'll try to clarify...